pytest_inmanta/plugin.py

```diff
@@ -739,6 +793,68 @@ def deploy(
         self.finalize_handler(h)
         return ctx
 
+    def deploy_all(self, run_as_root: bool = False) -> DeployResult:
+        """
+        Deploy all resources, in the correct order
+
+        This method handles skips and failures like the normal orchestrator
+
+        However, it can not handle Undefined resources.
+        """
+
+        def build_handler_and_context(
+            resource: Resource,
+        ) -> Tuple[Resource, ResourceHandler, HandlerContext]:
+            h = self.get_handler(resource, run_as_root)
+            assert h is not None
+            ctx = HandlerContext(resource)
+            return resource, h, ctx
+
+        all_contexts = {
+            str(rid): build_handler_and_context(resource)
+            for rid, resource in self.resources.items()
+        }
+
+        todo: List[str] = sorted(list(all_contexts.keys()))
+        order: List[str] = []
+
+        def topo_sort(doing: List[str], current: str):
+            # to be replace with graphlib.TopologicalSorter when we drop python 3.6 support
+            # Will not win a beauty contest, but it is stable and works
+            if current not in todo:
+                return
+            if current in doing:
+                raise Exception(f"Cycle detected: {doing}")
+
+            todo.remove(current)
+
+            new_doing = doing + [current]
+            for req in all_contexts[current][0].requires:
+                topo_sort(new_doing, str(req))
+            order.append(current)
+
+        while todo:
+            topo_sort([], todo[0])
+
+        for rid in order:
+            resource, h, ctx = all_contexts[rid]
+            skip = any(
+                all_contexts[str(dependency)][2].status != ResourceState.deployed
+                for dependency in resource.requires
+            )
+            if skip:
+                LOGGER.debug("Skipping %s", resource.id)
+                ctx.set_status(ResourceState.skipped)
+            else:
+                LOGGER.debug("Start executing %s", resource.id)
+                h.execute(ctx, resource)
+                LOGGER.debug("Done executing %s", resource.id)
+            self.finalize_context(ctx)
+            self.ctx = ctx
```

Comment:
So `self.ctx` references the `HandlerContext` of the resource that was deployed last. Isn't that a bit weird?

Reply:
This is always what is has been, but I admit it is a bit weird here. Do I drop it?

Reply:
I would. Leaving it in could only lead to bugs in test cases. Maybe we should set it to `None` explicitly after an invocation of `deploy_all()`

Reply:
ok!